### PR TITLE
Orca 447 add server config to run pre hooks

### PR DESCRIPTION
**Pull request by msarvar**

### CODE REVIEW

**Comment by nishkrishnan** [comment]
can you rebase? It should pull in the gh action that I pushed

File: server/events/workflow_hooks_command_runner.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	src := fmt.Sprintf("%s#%d", repoFullName, pullNum)
+	return w.Logger.NewLogger(src, true, w.Logger.GetLevel())
+}
+
+// logPanics logs and creates a comment on the pull request for panics.
+func (w *DefaultWorkflowHooksCommandRunner) logPanics(baseRepo models.Repo, pullNum int, logger logging.SimpleLogging) {
+	if err := recover(); err != nil {
+		stack := recovery.Stack(3)
+		logger.Err("PANIC: %s\n%s", err, stack)
+		if commentErr := w.VCSClient.CreateComment(
```
[the remark this answers is not in the document]
Comment (nishkrishnan) [reply server/events/workflow_hooks_command_runner.go:135]: Seems like a bad user experience to do that but don't want to deviate from the standard if that's set already.

File: server/events_controller.go [hunk 2]
```diff
@@ -435,6 +439,10 @@ func (e *EventsController) handleCommentEvent(w http.ResponseWriter, baseRepo mo
 		return
 	}
 
+	// TODO: run pre workflow hooks
+	// e.Logger.Info("running pre workflow hooks")
+	// e.WorkflowHooksCommandRunner.RunPreHooks(baseRepo, headRepo, pull, user)
+
```
Comment (nishkrishnan) [inline server/events_controller.go:445]: ?

File: server/server.go [hunk 3]
```diff
@@ -459,22 +470,23 @@ func NewServer(userConfig UserConfig, config Config) (*Server, error) {
...
-		IndexTemplate:       indexTemplate,
-		LockDetailTemplate:  lockTemplate,
-		SSLKeyFile:          userConfig.SSLKeyFile,
-		SSLCertFile:         userConfig.SSLCertFile,
-		Drainer:             drainer,
+		AtlantisVersion:            config.AtlantisVersion,
+		AtlantisURL:                parsedURL,
+		Router:                     underlyingRouter,
+		Port:                       userConfig.Port,
+		WorkflowHooksCommandRunner: workflowHooksCommandRunner,
```
Comment (nishkrishnan) [inline server/server.go:477]: PreWorkflowHooksCommandRunner*

**Review by nishkrishnan — COMMENTED** [review]
Nice job! Have some comments/questions

File: server/events/workflow_hooks_command_runner.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+//go:generate pegomock generate -m --use-experimental-model-gen --package mocks -o mocks/mock_workflows_hooks_command_runner.go WorkflowHooksCommandRunner
+
+type WorkflowHooksCommandRunner interface {
+	RunPreHooks(
+		baseRepo models.Repo,
+		headRepo models.Repo,
+		pull models.PullRequest,
+		user models.User,
+	)
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/workflow_hooks_command_runner.go:22]: Initially it was returning an object and error but `CommandRunner` interface doesn't have any return values. So decided to be consistent.

File: server/events/workflow_hooks_command_runner.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	ctx := models.WorkflowHookCommandContext{
+		BaseRepo: baseRepo,
+		HeadRepo: headRepo,
+		Log:      log,
+		Pull:     pull,
+		User:     user,
+		Verbose:  false,
+	}
+
+	result := w.runHooks(ctx, workflowHooks, repoDir)
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/workflow_hooks_command_runner.go:86]: Yeah, i was trying too hard to follow the atlantis implementation for run_steps. But it makes sense to reduce the wrapping.

File: server/events/workflow_hooks_command_runner.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/workflow_hooks_command_runner.go:135]: Yep, it is in the command_runner.go both `RunCommentCommand` and `RunAutoplanCommand` use it. What would alternative be? Potentially we can log and move on for pre_workflow_hooks.

File: server/events/workflow_hooks_command_result.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"strings"
+
+	"github.com/runatlantis/atlantis/server/events/models"
+)
+
+// WorkflowHooksCommandResult is the result of executing pre workflow hooks for a
+// repository.
+type WorkflowHooksCommandResult struct {
+	WorkflowHookResults []models.WorkflowHookResult
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/workflow_hooks_command_result.go:12]: You're right it is not needed here. I will remove it. It is used for command_runner where you have multiple projects running in parallel.

File: server/events/workflow_hooks_command_runner.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	defer w.logPanics(baseRepo, pull.Num, log)
+
+	log.Info("Running Pre Hooks for repo: ")
+
+	unlockFn, err := w.WorkingDirLocker.TryLock(baseRepo.FullName, pull.Num, DefaultWorkspace)
+	if err != nil {
+		log.Warn("workspace is locked")
+		return
+	}
+	log.Debug("got workspace lock")
```
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/workflow_hooks_command_runner.go:61]: this `log` is an instance of a logger and it uses `repo_name#pr_num` as a prefix. 
This is an example from my local
```
2020/10/19 16:43:53-0700 [DBUG] pre_workflow_hooks_command_runner.go:61 msarvar/atlantis-local-dev#28: Got workspace lock
```

[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/workflow_hooks_command_runner.go:61]: I can't do better than repo_name and PR since I don't have root information at this point.

File: server/events/workflow_hooks_command_runner.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nishkrishnan) [reply server/events/workflow_hooks_command_runner.go:135]: It's a good question.

i think in this case, the hooks are going to be only configured server side right? So customers shouldn't really know about it. Hence I would lean towards something like we just log an error and a stat (and alarm on this internally) and just move on.

In our case, there will probably be some downstream error about missing an atlantis configuration file so it's probably ok. In the general case, I'm not sure what people would use this for, but might be safer to not panic on failure in case it's something non-critical.

What do you think?

File: server/events/workflow_hooks_command_runner.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (nishkrishnan) [reply server/events/workflow_hooks_command_runner.go:22]: Hmm I see yeah ok we can probably be consistent then.

File: server/events/yaml/raw/global_cfg.go [hunk 8]
```diff
@@ -18,11 +18,12 @@ type GlobalCfg struct {
...
 // Repo is the raw schema for repos in the server-side repo config.
 type Repo struct {
-	ID                   string   `yaml:"id" json:"id"`
-	ApplyRequirements    []string `yaml:"apply_requirements" json:"apply_requirements"`
-	Workflow             *string  `yaml:"workflow,omitempty" json:"workflow,omitempty"`
-	AllowedOverrides     []string `yaml:"allowed_overrides" json:"allowed_overrides"`
-	AllowCustomWorkflows *bool    `yaml:"allow_custom_workflows,omitempty" json:"allow_custom_workflows,omitempty"`
+	ID                   string         `yaml:"id" json:"id"`
+	ApplyRequirements    []string       `yaml:"apply_requirements" json:"apply_requirements"`
+	PreWorkflowHooks        []PreWorkflowHook `yaml:"pre_workflow_hooks" json:"pre_workflow_hooks"`
```
Comment (nishkrishnan) [inline server/events/yaml/raw/global_cfg.go]: I'm assuming we don't support this in the repo config currently?

File: server/events/workflow_hooks_command_runner.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (msarvar) [reply server/events/workflow_hooks_command_runner.go]: I think that should work for our use case.

**Review by nishkrishnan — APPROVED** [review]
🚢

File: server/events/yaml/raw/global_cfg.go (hunk 8 above)
Comment (msarvar) [reply server/events/yaml/raw/global_cfg.go]: Not sure what would be the use case for repo config? For repo config there is a custom run commands.

**Comment by msarvar** [comment]
💨
